### PR TITLE
Remove helper frames

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
Convert GC function into QCall

Port CustomAttribute data parsing into C#.

### CODE REVIEW

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs [hunk 1]
```diff
@@ -704,83 +672,439 @@ internal enum CustomAttributeEncoding : int
...
+
+            private ReadOnlySpan<byte> PeekData(int size)
+            {
+                ArgumentOutOfRangeException.ThrowIfLessThan(_blob.Length - _curr, size);
+                return _blob.Slice(_curr, size);
+            }
+
+            private ReadOnlySpan<byte> ReadData(int size)
+            {
+                var tmp = PeekData(size);
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs:937]:
```suggestion
                ReadOnlySpan<byte> tmp = PeekData(size);
```
This is invalid use of `var` per https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs [hunk 2]
```diff
@@ -704,83 +672,439 @@ internal enum CustomAttributeEncoding : int
...
+                        arg.ArrayValue = new CustomAttributeEncodedArgument[len];
+                        for (int i = 0; i < len; ++i)
+                        {
+                            arg.ArrayValue[i] = ParseCustomAttributeValue(ref parser, attributeType, module);
+                        }
+                    }
+                    break;
+                }
+                default:
+                    throw new ArgumentOutOfRangeException();
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs:870]: I think these and similar places should be `BadImageFormatException`.

**Comment by lambdageek** [comment]
How do you avoid chicken-and-egg problems? does this custom attribute parser get used when you are loading some class or JITing a method?  What if there was `[InlineArray]` data in `RuntimeCustomAttributeData.cs` that the parser itself depended on, for example?

**Comment by AaronRobinsonMSFT** [comment]
@lambdageek I didn't remove the parser itself, just the infrastructure around calling the native parser from managed code. This was very old code and the process was rather complicated.

1) Partially populate, using Reflection, managed types with details for the signature parsing
2) Pass those managed types down into native via an FCall
3) Convert the data in the managed types into the native types
4) Parsing the signature blob using the details in the native types
5) Place the data from the native types into the managed types
6) Move back into managed code and lazily convert the raw details into public surface types

This PR simply collapses 2 - 5 and just does it all in C#.

File: src/libraries/System.Runtime/tests/System.Runtime.Tests/System/AttributeTests.cs [hunk 3]
```diff
@@ -167,7 +167,7 @@ public static void StringArgument_InvalidCodeUnits_FallbackUsed()
 
             string stringArg = cad.ConstructorArguments[0].Value as string;
             Assert.NotNull(stringArg);
-            Assert.Equal("\uFFFD\uFFFD", stringArg);
+            Assert.Equal("\uFFFD\uFFFD\uFFFD", stringArg);
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/System.Runtime.Tests/System/AttributeTests.cs:170]: This is failing on Mono now.

File: src/libraries/System.Private.CoreLib/src/Resources/Strings.resx [hunk 4]
```diff
@@ -1,17 +1,17 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <root>
-  <!-- 
-    Microsoft ResX Schema 
-    
+  <!--
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:3]: Nit: Revert whitespace-only changes in this file

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs [hunk 5]
```diff
@@ -704,83 +672,439 @@ internal enum CustomAttributeEncoding : int
...
-        public CustomAttributeEncoding EncodedType => m_encodedType;
-        public CustomAttributeEncoding EncodedEnumType => m_encodedEnumType;
-        public CustomAttributeEncoding EncodedArrayType => m_encodedArrayType;
-        public string? EnumName => m_enumName;
+        public Type? EnumType => enumType;
+
+        public static CustomAttributeType CreateFromType(RuntimeType parameterType)
+        {
+            Debug.Assert(parameterType is not null);
+            var encodedType = RuntimeCustomAttributeData.TypeToCustomAttributeEncoding(parameterType);
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs:1088]: Nit: This is not allowed use of var per repo coding convention.

File: src/libraries/System.Runtime/tests/System.Runtime.Tests/System/AttributeTests.cs (hunk 3 above)
Comment (lambdageek) [reply src/libraries/System.Runtime/tests/System.Runtime.Tests/System/AttributeTests.cs:170]: Might be related to https://github.com/dotnet/runtime/commit/beeb5e009fec205e334faaddc98944b522c2cd03#diff-9bd9d0e46241d07dc62110246c89c9f056333eff4b0b089551780a625c871d0d ?

https://github.com/dotnet/runtime/blob/135fec006e727a31763271984cd712f1659ccbd3/src/mono/mono/metadata/custom-attrs.c#L414-L419

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime/tests/System.Runtime.Tests/System/AttributeTests.cs:170]: Thanks for the tip @lambdageek, that does look promising.

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime/tests/System.Runtime.Tests/System/AttributeTests.cs:170]: Sad. yes, this is the offending call, but it is doing the right thing and calling into the native utf8.c implementation. The native implementation matches the Win32 `MultiByteToWideChar()` API, yay, but fails differently than the managed UTF-8 parser, the WTF-8 behavior that is being alluded to here is the issue.

I am using `Encoding.UTF8.GetString()` in the managed impl so perhaps I need to use something else instead. I will start there.

Comment (jkotas) [reply src/libraries/System.Runtime/tests/System.Runtime.Tests/System/AttributeTests.cs:170]: I do not think we care about the exact failure more for invalid UTF8 strings in the binaries. The behavior is not tested in general and I do not expect it to be consistent.

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime/tests/System.Runtime.Tests/System/AttributeTests.cs:170]: Yay, thank you for letting me know.

File: src/coreclr/vm/comutilnative.cpp [hunk 6]
```diff
@@ -801,26 +801,31 @@ extern "C" int QCALLTYPE GCInterface_EndNoGCRegion()
 **Arguments: args->handle -- the OBJECTHANDLE to the object which we're locating.
 **Exceptions: ArgumentException if handle points to an object which is not accessible.
 ==============================================================================*/
-FCIMPL1(int, GCInterface::GetGenerationWR, LPVOID handle)
+extern "C" int QCALLTYPE GCInterface_GetGenerationWR(LPVOID handle)
```
Comment (jkotas) [inline src/coreclr/vm/comutilnative.cpp:804]: This QCall can be deleted and moved to C#. Something like:

```
        public static int GetGeneration(WeakReference wo)
        {
            // Note - This throws an NRE if given a null weak reference.
            object? obj = GCHandle.InternalGet(wo.WeakHandle);
            KeepAlive(wo);

            if (obj == null)
            {
                throw new ArgumentNullException(nameof(wo));
            }

            return GetGeneration(obj);
        }
```

Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/comutilnative.cpp:804]: So I was going to that actually, but I didn't want to extend the lifetime of the object by extracting it from the `WeakReference`. Does that not matter much in this case?

Comment (jkotas) [reply src/coreclr/vm/comutilnative.cpp:804]: I do not think it matters in this case.

This API is useless, it has 0 known uses https://apisof.net/catalog/364d5dc6-3461-d876-bdc8-8a51a38b6606.

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs [hunk 7]
```diff
@@ -704,83 +672,439 @@ internal enum CustomAttributeEncoding : int
...
-        private readonly CustomAttributeEncodedArgument m_encodedArgument;
+        public CustomAttributeType CustomAttributeType { get; private set; }
+        public PrimitiveValue PrimitiveValue { get; set; }
+        public CustomAttributeEncodedArgument[]? ArrayValue { get; set; }
+        public string? StringValue { get; set; }
 
-        public CustomAttributeNamedParameter(string argumentName, CustomAttributeEncoding fieldOrProperty, CustomAttributeType type)
+        private static void ParseCtorArgs(
+            ref CustomAttributeDataParser parser,
+            ref CustomAttributeCtorParameter[] customAttributeCtorParameters,
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs:705]:
```suggestion
            CustomAttributeCtorParameter[] customAttributeCtorParameters,
```
Does not need to be `ref`?

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs [hunk 8]
```diff
@@ -704,83 +672,439 @@ internal enum CustomAttributeEncoding : int
...
+                p.EncodedArgument = ParseCustomAttributeValue(
+                    ref parser,
+                    p.CustomAttributeType,
+                    module);
+            }
+        }
+
+        private static void ParseNamedArgs(
+            ref CustomAttributeDataParser parser,
+            ref CustomAttributeNamedParameter[] customAttributeNamedParameters,
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs:719]:
```suggestion
            CustomAttributeNamedParameter[] customAttributeNamedParameters,
```

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs [hunk 9]
```diff
@@ -704,83 +672,439 @@ internal enum CustomAttributeEncoding : int
...
+                    break;
+                case CustomAttributeEncoding.String:
+                case CustomAttributeEncoding.Type:
+                    arg.StringValue = parser.GetString();
+                    break;
+                case CustomAttributeEncoding.Array:
+                {
+                    arg.ArrayValue = null;
+                    int len = parser.GetI4();
+                    if (len != unchecked((int)0xffffffff)) // indicates array is null - ECMA-335 II.23.3.
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs:854]:
```suggestion
                    if (len != -1) // indicates array is null - ECMA-335 II.23.3.
```

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs [hunk 10]
```diff
@@ -704,83 +672,439 @@ internal enum CustomAttributeEncoding : int
...
+                unsafe
+                {
+                    _blob = new ReadOnlySpan<byte>((void*)attributeBlob.Signature, attributeBlob.Length);
+                }
+                _curr = 0;
+            }
+
+            private ReadOnlySpan<byte> PeekData(int size)
+            {
+                ArgumentOutOfRangeException.ThrowIfLessThan(_blob.Length - _curr, size);
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs:931]:
```suggestion
```
We can just let `Slice` to throw the exception

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs [hunk 11]
```diff
@@ -704,83 +672,439 @@ internal enum CustomAttributeEncoding : int
...
-    internal readonly struct CustomAttributeType
+    internal sealed class CustomAttributeNamedParameter(MemberInfo memberInfo, CustomAttributeEncoding fieldOrProperty, CustomAttributeType type)
     {
+        public MemberInfo MemberInfo => memberInfo;
+        public CustomAttributeType CustomAttributeType => type;
+        public CustomAttributeEncoding FieldOrProperty => fieldOrProperty;
+        public CustomAttributeEncodedArgument? EncodedArgument { get; set; }
+    }
+
+    internal sealed class CustomAttributeType(
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs:1072]: @stephentoub Do we have a preferred naming style for primary constructors? Do we want the instance fields that they create to have `_` prefix?

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs [hunk 12]
```diff
@@ -676,23 +647,20 @@ internal enum CustomAttributeEncoding : int
...
+        [FieldOffset(0)]
+        public long Byte8;
+    }
+
+    internal sealed class CustomAttributeEncodedArgument
+    {
+        internal static void ParseAttributeArguments(
+            ConstArray attributeBlob,
             ref CustomAttributeCtorParameter[] customAttributeCtorParameters,
             ref CustomAttributeNamedParameter[] customAttributeNamedParameters,
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs:665]:
```suggestion
            CustomAttributeCtorParameter[] customAttributeCtorParameters,
            CustomAttributeNamedParameter[] customAttributeNamedParameters,
```

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs (hunk 11 above)
Comment (stephentoub) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs:1072]: > Do we have a preferred naming style for primary constructors? Do we want the instance fields that they create to have _ prefix?

The same as for any normal constructor.

It's hard to put into guidelines, but I have a preference for us only using primary constructors when either:
a) The body of the class is so small that it's obvious just by looking that a primary ctor parameter is being used, or
b) Having the class declare fields that are initialized to the primary ctor parameter, with the rest of the class then referring to those fields. This lets you make them readonly, for example.

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs [hunk 13]
```diff
@@ -704,83 +672,439 @@ internal enum CustomAttributeEncoding : int
...
 
-    [StructLayout(LayoutKind.Auto)]
-    internal readonly struct CustomAttributeNamedParameter
-    {
-        private readonly string m_argumentName;
-        private readonly CustomAttributeEncoding m_fieldOrProperty;
-        private readonly CustomAttributeEncoding m_padding;
-        private readonly CustomAttributeType m_type;
-        private readonly CustomAttributeEncodedArgument m_encodedArgument;
+        public CustomAttributeType CustomAttributeType { get; private set; }
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs:698]:
```suggestion
        public CustomAttributeType CustomAttributeType { get; }
```

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs [hunk 14]
```diff
@@ -704,83 +672,439 @@ internal enum CustomAttributeEncoding : int
...
+            {
+                // We cannot determine the underlying type without loading the enum.
+                string? enumTypeMaybe = parser.GetString();
+                if (enumTypeMaybe is null)
+                {
+                    throw new BadImageFormatException();
+                }
+
+                enumType = TypeNameParser.GetTypeReferencedByCustomAttribute(enumTypeMaybe, module);
+                if (enumType is null || !enumType.IsEnum)
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs:901]:
```suggestion
                if (!enumType.IsEnum)
```
This cannot return null

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs [hunk 15]
```diff
@@ -704,83 +672,439 @@ internal enum CustomAttributeEncoding : int
...
+            return arg;
+        }
+
+        private static CustomAttributeType ParseCustomAttributeType(ref CustomAttributeDataParser parser, RuntimeModule module)
         {
-            m_type = type;
-            m_encodedArgument = default;
+            CustomAttributeEncoding arrayTag = CustomAttributeEncoding.Undefined;
+            CustomAttributeEncoding enumTag = CustomAttributeEncoding.Undefined;
+            Type? enumType = null;
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs:880]:
```suggestion
```

File: src/libraries/System.Runtime/tests/System.Runtime.Tests/System/AttributeTests.cs [hunk 16]
```diff
@@ -167,7 +167,15 @@ public static void StringArgument_InvalidCodeUnits_FallbackUsed()
 
             string stringArg = cad.ConstructorArguments[0].Value as string;
             Assert.NotNull(stringArg);
-            Assert.Equal("\uFFFD\uFFFD", stringArg);
+
+            // Validate that each character is 'invalid'.
+            // The runtimes are inconsistent with respect to conversion
+            // failure modes with respect to length so we just validate
+            // each character.
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime/tests/System.Runtime.Tests/System/AttributeTests.cs:174]:
```suggestion
            // The runtimes are inconsistent with respect to conversion
            // failure modes so we just validate each character.
```

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs (hunk 15 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeCustomAttributeData.cs:880]: This is used.

**Review by jkotas — APPROVED** [review]
Thank you!

**Pull request merged**